src/xiaomi_ble/cloud.py

```diff
@@ -0,0 +1,452 @@
+        r.encrypt(bytes(1024))
+        return base64.b64encode(r.encrypt(payload.encode())).decode()
+
+    @staticmethod
+    def decrypt_rc4(password: str, payload: str) -> str:
+        """Decrypt a piece of data."""
+        r = ARC4.new(base64.b64decode(password))
+        r.encrypt(bytes(1024))
+        return r.encrypt(base64.b64decode(payload))
+
+
+class XiaomiCloudTokenFetch:
+
+    def __init__(
+        self, username: str, password: str, session: aiohttp.ClientSession
+    ) -> None:
+        """Initialize the Xiaomi Cloud API."""
+        self._username = username
+        self._password = password
+        self._session = session
+
+    async def get_device_info(
+        self, mac: str, servers: list[str] = SERVERS
+    ) -> XiaomiCloudBLEDevice | None:
+        """Get the token for a given MAC address."""
+        formatted_mac = format_mac_upper(mac)
+        connector = XiaomiCloudConnector(self._username, self._password, self._session)
+        await connector.login()
+        assert connector.userId is not None
+        homes: dict[str, str] = {}
+
+        for server in servers:
+            home_info = await connector.get_homes(server)
+            if home_info is not None:
+                for h in home_info["result"]["homelist"]:
+                    homes[h["id"]] = connector.userId
+
+            dev_cnt = await connector.get_dev_cnt(server)
+            if dev_cnt is not None:
+                for h in dev_cnt["result"]["share"]["share_family"]:
+                    homes[h["home_id"]] = h["home_owner"]
+
+            for home_id, owner_id in homes.items():
+                devices = await connector.get_devices(server, home_id, owner_id)
+                if (
+                    devices is None
+                    or not devices["result"]
+                    or not devices["result"]["device_info"]
+                ):
+                    continue
+
+                device_info: list[dict[str, Any]] = devices["result"]["device_info"]
+                for device in device_info:
+                    if (
+                        device["mac"] != formatted_mac
+                        or "did" not in device
+                        or "blt" not in device["did"]
+                    ):
+                        continue
+                    key_result = await connector.get_beaconkey(server, device["did"])
+                    if (
+                        key_result
+                        and (result := key_result.get("result"))
+                        and (beacon_key := result.get("beaconkey"))
+                    ):
+                        _LOGGER.debug(
+                            "Found beacon key for %s: %s (%s)",
+                            formatted_mac,
+                            beacon_key,
+                            device,
+                        )
+                        return XiaomiCloudBLEDevice(
+                            device["name"], device["mac"], beacon_key
```

Comment:
Should we postfix with `FFFFFFFF` if its 24 bytes instead of 32?

https://github.com/Bluetooth-Devices/xiaomi-ble/issues/96#issuecomment-2412050140

Reply:
Hmmm, not sure. Only if it is using MiBeacon v4/v5 and you have a 12 bytes key (24characters) . MiBeancon V2/V3 is using a 12 bytes key, so, not in that case

Reply:
Let's leave it as-is, we can wait for someone to bring it up